Allow graph fuser to move chunks past multiple nodes.

zou3519 opened the pull request:
Fixes #12290. Also speeds up JIT LSTM forward pass from 8.8ms to 7.8ms; previously, each JIT lstm cell used 2 fused kernels. Now, it only uses one fused kernel (which is how many kernels cudnn uses).

Explanation:

Let f, g, h be fusible ops.
```
x = f(v, w)
z = g(x, y)
a, b = chunk(z)
c = h(a, b)
```
becomes (before this PR):
```
x = f(v, w)
x', y' = broadcast_tensors([x, y])
ax, bx = chunk(x')
ay, by = chunk(y')
a = g(ax, ay)
b = g(bx, by)
c = h(a, b)
```
The graph fuser then puts g, g, and h into one FusionGroup and is unable
to move `x = f(v, w)` into the FusionGroup.

This PR lets the graph fuser move `x = f(v, w)` into the FusionGroup.
It does this by abstracting the broadcast_tensors + multiple chunk nodes
into one intermediate `prim::BroadcastingChunk[chunks, dim]` node.

A `BroadcastingChunk[chunks, dim](*inputs)` node is equivalent to:
- broadcasting all of *inputs
- chunk-ing each broadcasted input into `chunks` chunks along dim `dim`.

Abstracting the broadcasting chunk behavior away, it is now a lot easier
for the graph fuser to move (broadcast + chunk) past an operation. After
this PR, the above graph becomes:
```
x = f(v, w)
ax, bx, ay, by = BroadcastingChunk(x, y)
a = g(ax, ay)
b = g(bx, by)
c = h(a, b)
```
Now, to move `x = f(v, w)` after the BroadcastingChunk, one just needs
to add f's operands to the BroadcastingChunk:
```
ay, by, av, bv, aw, bw = BroadcastingChunk(y, v, w)
ax = f(av, aw)
by = f(bv, bw)
a = g(ax, ay)
b = g(bx, by)
c = h(a, b)
```

cc @apaszke @mruberry @zdevito

zou3519 committed:
Allow graph fuser to move chunks past multiple nodes.

Fixes #12290. Also speeds up JIT LSTM forward pass from 8.8ms to 7.8ms.

Explanation:

Let f, g, h be fusible ops.
```
x = f(v, w)
z = g(x, y)
a, b = chunk(z)
c = h(a, b)
```
becomes (before this PR):
```
x = f(v, w)
x', y' = broadcast_tensors([x, y])
ax, bx = chunk(x')
ay, by = chunk(y')
a = g(ax, ay)
b = g(bx, by)
c = h(a, b)
```
The graph fuser then puts g, g, and h into one FusionGroup and is unable
to move `x = f(v, w)` into the FusionGroup.

This PR lets the graph fuser move `x = f(v, w)` into the FusionGroup.
It does this by abstracting the broadcast_tensors + multiple chunk nodes
into one intermediate prim::BroadcastingChunk[chunks, dim] node.

A BroadcastingChunk[chunks, dim](*inputs) node is equivalent to:
- broadcasting all of *inputs
- chunk-ing each broadcasted input into `chunks` chunks along dim `dim`.

Abstracting the broadcasting chunk behavior away, it is now a lot easier
for the graph fuser to move (broadcast + chunk) past an operation. After
this PR, the above graph becomes:
```
x = f(v, w)
ax, bx, ay, by = BroadcastingChunk(x, y)
a = g(ax, ay)
b = g(bx, by)
c = h(a, b)
```
Now, to move `x = f(v, w)` after the BroadcastingChunk, one just needs
to add f's operands to the BroadcastingChunk:
```
ay, by, av, bv, aw, bw = BroadcastingChunk(y, v, w)
ax = f(av, aw)
by = f(bv, bw)
a = g(ax, ay)
b = g(bx, by)
c = h(a, b)
```

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -523,23 +599,25 @@ struct GraphFuser {
-    Node * insertion_point = chunk;
+    std::vector<Value*> producer_chunk_outputs;
+    for (size_t i = 0; i < nchunks; i++) {
+      producer_chunk_outputs.push_back(bchunk->output(4 * producer_index + i));
```
What does this hardcoded `4` correspond to? Shouldn't this be `nchunks`?

zou3519 replied:
Yes, thank you for the catch!

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -551,23 +629,18 @@ struct GraphFuser {
-          Value * input_chunk_sel = input_chunk->addOutput();
+      for (auto chunk_sel : producer_chunk_outputs) {
+          Value * input_chunk_sel = bchunk->addOutput();
           input_chunk_sel->setType(chunk_sel->type());
```
FWIW this is not true if we were operating on `CompleteTensorType`s (it might get broadcast), but not sure how important is that

zou3519 replied:
Good point. I'll add an assert on the type for now

zou3519 replied:
Actually, I think this is valid for CompleteTensorType (do you have an example where it isn't?). All outputs of the `BroadcastingChunk` have the same shape; consider the following operations:
```
z = x * y
z0, z1 = BroadcastingChunk(z)
```
We only add new inputs to the BroadcastingChunk if they're inputs to nodes that are producers for the BroadcastingChunk. In this case, we can add `x`, `y` as inputs to the bchunk while removing `z`. The new outputs `x0, x1, y0, y1` will still have the same shape as `z0` and `z1` because the bchunk is equivalent to broadcasting `x, y` to be the same shape as `z` and then performing chunking.

apaszke replied:
You're right!

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -523,23 +599,25 @@ struct GraphFuser {
+      bchunk = promoteChunkToBroadcastingChunk(chunk);
+    }
+    size_t nchunks = bchunk->i(attr::chunks);
+    WithInsertPoint guard(bchunk);
```
Shouldn't we take the node *after* the `bchunk`? IIRC this means that all news nodes will be inserted before `bchunk`, while you really want to repeat the operations outside of it.

apaszke commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -551,23 +629,18 @@ struct GraphFuser {
-      input_chunk->i_(attr::dim, chunk->i(attr::dim));
-      insertAt(&insertion_point, input_chunk);
-
+      bchunk->addInput(input);
```
nit: it would be good to deduplicate inputs (it's really easy now!)

apaszke reviewed (approved):
Looks awesome :clap:

Would be good to expand our test suite a bit in this case. Ultimately we want all ways to parenthesize this expression (possibly with permutation) `x.mm(w_ih) + h.mm(w_hh) + b_ih + b_hh` to get fused correctly into some operations following the chunk, but we're only checking one at the moment.

I want to think about this PR a bit more, but I'm pretty confident that it's good and can land.

zou3519 committed:
Address comments:

- Minor fixes
- Deduplicate inputs to BroadcastingChunk
- test deduplication of inputs
- test permutations of gates = x.mm(w_ih.t()) + hx.mm(w_hh.t()) + b_ih +
  b_hh and assert all of them result in one FusionGroup (as opposed to
  multiple)

zou3519 committed:
Rebase; add BroadcastingChunk to alias analysis

zou3519 commented:
Windows test looks flaky (I see it on other PRs as well).
